Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,221 @@
+  let fixIts: [FixIt]  // TODO: How to create an abstract model for this?
+
+  /// Initializes a new `DiagnosticDescriptor`.
+  ///
+  /// - Parameters:
+  ///   - locationMarker: The marker pointing to location in source code.
+  ///   - id: The message ID of the diagnostic.
+  ///   - message: The textual message to display for the diagnostic.
+  ///   - severity: The severity level of the diagnostic. Default is `.error`.
+  ///   - highlight: The syntax elements to be highlighted. Default is an empty array.
+  ///   - noteDescriptors: An array of note descriptors for additional context. Default is an empty array.
+  ///   - fixIts: An array of Fix-It descriptors for quick fixes. Default is an empty array.
+  init(
+    locationMarker: LocationMarker,
+    id: MessageID = MessageID(domain: "test", id: "conjured"),
+    message: String,
+    severity: DiagnosticSeverity = .error,
+    highlight: [Syntax] = [],
+    noteDescriptors: [NoteDescriptor] = [],
+    fixIts: [FixIt] = []
+  ) {
+    self.locationMarker = locationMarker
+    self.id = id
+    self.message = message
+    self.severity = severity
+    self.highlight = highlight
+    self.noteDescriptors = noteDescriptors
+    self.fixIts = fixIts
+  }
+
+  /// Creates a ``Diagnostic`` instance from a given ``DiagnosticDescriptor``, syntax tree, and location markers.
+  ///
+  /// - Parameters:
+  ///   - tree: The syntax tree where the diagnostic is rooted.
+  ///   - markers: A dictionary mapping location markers to their respective offsets in the source code.
+  ///   - file: The file where the test is located, used for reporting failures.
+  ///   - line: The line where the test is located, used for reporting failures.
+  ///
+  /// - Returns: A ``Diagnostic`` instance populated with details from the ``DiagnosticDescriptor``, or `nil` if it fails.
+  func createDiagnostic(
+    inSyntaxTree tree: some SyntaxProtocol,
+    usingLocationMarkers markers: [LocationMarker: Int],
+    file: StaticString = #file,
+    line: UInt = #line
+  ) -> Diagnostic? {
+    func node(at marker: LocationMarker) -> Syntax? {
+      guard let markedOffset = markers[marker] else {
+        XCTFail("Marker \(marker) not found in the marked source", file: file, line: line)
+        return nil
+      }
+      let markedPosition = AbsolutePosition(utf8Offset: markedOffset)
+      guard let token = tree.token(at: markedPosition) else {
+        XCTFail("Node not found at marker \(marker)", file: file, line: line)
+        return nil
+      }
+      return Syntax(token)
+    }
+
+    guard let diagnosticNode = node(at: self.locationMarker) else { return nil }
+
+    var notes = [Note]()
+    for noteDescriptor in self.noteDescriptors {
+      guard let noteNode = node(at: noteDescriptor.locationMarker) else { continue }
+
+      let note = Note(
+        node: noteNode,
+        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
+      )
+      notes.append(note)
+    }
```

Comment:
I think using `map` here might make it a little bit more swifty + easier to understand, especially if `node(at:)` throws on failure.

Reply:
My idea was to avoid stoping the assertion process if just one of the nodes couldn't be found. That's why I've used a `for` loop instead of `map`, as I'm employing `continue` to skip to the next iteration. However, I believe we could also use `compactMap` as an alternative.

Reply:
Changed to `map` – done.